test/fuzz/utils.hpp

```diff
@@ -0,0 +1,218 @@
+// Copyright (C) 2023 Intel Corporation
+// Part of the Unified-Runtime Project, under the Apache License v2.0 with LLVM Exceptions.
+// See LICENSE.TXT
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <fstream>
+#include <fuzzer/FuzzedDataProvider.h>
+#include <iostream>
+#include <iterator>
+#include <map>
+#include <memory>
+#include <vector>
+
+namespace fuzz {
+
+enum FuzzerAPICall : uint8_t {
+    UR_PLATFORM_GET,
+    UR_DEVICE_GET,
+    UR_DEVICE_RELEASE,
+    UR_CONTEXT_CREATE,
+    UR_CONTEXT_RELEASE,
+    UR_USM_POOL_CREATE_HOST,
+    UR_USM_POOL_CREATE_DEVICE,
+    UR_USM_POOL_RELEASE_HOST,
+    UR_USM_POOL_RELEASE_DEVICE,
+    UR_USM_HOST_ALLOC_POOL,
+    UR_USM_HOST_ALLOC_NO_POOL,
+    UR_USM_DEVICE_ALLOC_POOL,
+    UR_USM_DEVICE_ALLOC_NO_POOL,
+    UR_USM_FREE_HOST_POOL,
+    UR_USM_FREE_HOST_NO_POOL,
+    UR_USM_FREE_DEVICE_POOL,
+    UR_USM_FREE_DEVICE_NO_POOL,
+    UR_PROGRAM_CREATE_WITH_IL,
+    kMaxValue = UR_PROGRAM_CREATE_WITH_IL,
+};
+
+struct Alloc {
+    ur_context_handle_t context;
+    void *ptr;
+
+    Alloc(ur_context_handle_t context, void *ptr)
+        : context(context), ptr(ptr) {}
+    ~Alloc() { urUSMFree(context, ptr); }
+};
+using Allocs = std::vector<std::unique_ptr<Alloc>>;
+
+struct Pool {
+    ur_usm_pool_handle_t handle;
+    Allocs allocs;
+
+    Pool(ur_usm_pool_handle_t handle) : handle(handle) {}
+    ~Pool() {
+        allocs.clear();
+        urUSMPoolRelease(handle);
+    }
+};
+using Pools = std::vector<std::unique_ptr<Pool>>;
+
+struct Context {
+    ur_context_handle_t handle;
+    Pools host_pools;
+    Pools device_pools;
+    Allocs no_pool_host_allocs;
+    Allocs no_pool_device_allocs;
+
+    Context(ur_context_handle_t handle) : handle(handle) {}
+    ~Context() {
+        host_pools.clear();
+        device_pools.clear();
+        no_pool_host_allocs.clear();
+        no_pool_device_allocs.clear();
+        urContextRelease(handle);
+    }
+};
+using Contexts = std::vector<std::unique_ptr<Context>>;
+
+struct TestState {
+    static constexpr uint32_t num_entries = 1;
+
+    std::unique_ptr<FuzzedDataProvider> data_provider;
+
+    std::vector<ur_adapter_handle_t> adapters;
+    std::vector<ur_platform_handle_t> platforms;
+    std::vector<ur_device_handle_t> devices;
+    Contexts contexts;
+    ur_device_type_t device_type = UR_DEVICE_TYPE_ALL;
+
+    uint32_t num_adapters;
+    uint32_t num_platforms;
+    uint32_t num_devices;
+
+    uint8_t platform_num;
+    uint8_t device_num;
+    uint8_t context_num;
+
+    TestState(std::unique_ptr<FuzzedDataProvider> data_provider)
+        : data_provider(std::move(data_provider)) {}
+
+    template <typename IntType> int get_next_input_data(IntType *data) {
+        if (data_provider->remaining_bytes() < sizeof(IntType)) {
```

Comment:
I'm wondering if checking for remaining_bytes here and returning -1 makes sense. Why not just always call `consume*`, even if there are no bytes left? From the documentation, it will just return the same value every time (e.g. 0 for int). 

I don't know which approach is better for coverage but it certainly seems simpler as we don't need to deal with `-1`.

But feel free to ignore this comment. We can also experiment later with what gives better results.

Reply:
This way we would eventually get stuck with multiple calls to urPlatformGet(), as it's the first item in enum representing 0. I prefer to have it this way and terminate the iteration on insufficient data.